library/OpenTelemetry/AWSXRay/TraceInfo.hs

```diff
@@ -0,0 +1,120 @@
+--
+module OpenTelemetry.AWSXRay.TraceInfo
+  ( TraceInfo(..)
+  , fromXRayHeader
+  , toXRayHeader
+  ) where
+
+import Prelude
+
+import Control.Error.Util (note)
+import Data.Bifunctor (first)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BS8
+import qualified OpenTelemetry.AWSXRay.Baggage as Baggage
+import OpenTelemetry.Baggage (Baggage)
+import OpenTelemetry.Trace (SpanContext(..))
+import OpenTelemetry.Trace.Core
+  (defaultTraceFlags, isSampled, setSampled, unsetSampled)
+import OpenTelemetry.Trace.Id
+  ( Base(..)
+  , baseEncodedToSpanId
+  , baseEncodedToTraceId
+  , spanIdBaseEncodedByteString
+  , traceIdBaseEncodedByteString
+  )
+import qualified OpenTelemetry.Trace.TraceState as TS
+
+-- | The data to read/write from the @X-Amzn-TraceId@ header
+data TraceInfo = TraceInfo
+  { spanContext :: SpanContext
+  , baggage :: Maybe Baggage
+  }
+  deriving stock Show
+
+fromXRayHeader :: ByteString -> Either String TraceInfo
+fromXRayHeader bs = do
+  kv <- bsToKeyValues bs
+
+  root <- note "Root not present" $ lookup "Root" kv
+  traceId <- case bsSplitOn '-' root of
+    ["1", epoch, unique] -> do
+      -- epoch can be "at most" 8 hex and we should zero-pad the
+      -- short values. NB. This relies on hex being 1 char per char
+      let
+        epochUnique = bsLeftPad 8 '0' epoch <> unique
+        errorPrefix =
+          "Root epoch+unique ("
+            <> show epochUnique
+            <> ") is not a valid TraceId"
+      prefix errorPrefix $ baseEncodedToTraceId Base16 epochUnique
+    _ -> Left "Splitting on - did not produce exactly 3 parts"
+
+  parent <- note "Parent not present" $ lookup "Parent" kv
+  spanId <- prefix "Parent is not a valid SpanId"
+    $ baseEncodedToSpanId Base16 parent
+
+  let
+    sampled = (== Just "1") $ lookup "Sampled" kv
+
+    traceFlags =
+      (if sampled then setSampled else unsetSampled) defaultTraceFlags
+
+    baggage = Baggage.decode kv
+
+  pure $ TraceInfo
+    { spanContext = SpanContext
+      { traceFlags
+      , traceId
+      , spanId
+      , isRemote = True
+      , traceState = TS.empty
+      }
+    , baggage
+    }
+
+toXRayHeader :: TraceInfo -> ByteString
+toXRayHeader TraceInfo { spanContext, baggage } =
+  bsFromKeyValues
+    $ [ ("Root", "1-" <> epoch <> "-" <> unique)
+      , ("Parent", spanIdBaseEncodedByteString Base16 spanId)
+      , ("Sampled", if isSampled traceFlags then "1" else "0")
+      ]
+    <> maybe [] Baggage.encode baggage
+ where
+  SpanContext { traceId, spanId, traceFlags } = spanContext
+  (epoch, unique) = BS.splitAt 8 $ traceIdBaseEncodedByteString Base16 traceId
+
+prefix :: String -> Either String a -> Either String a
+prefix p = first (\e -> p <> ": " <> e)
+
+bsToKeyValues :: ByteString -> Either String [(ByteString, ByteString)]
+bsToKeyValues = traverse go . bsSplitOn ';'
+ where
+  go bs = case bsSplitOn '=' bs of
+    k : vs -> Right (k, mconcat vs)
```

Comment:
Could guard against `BS.null k` here. :shrug:

Reply:
If it makes an "odd gotcha" for consumers into a specific error I'm for it.

Though I guess where do you draw the line? Check it's not only whitespace? Check it's not weird characters that xray doesn't support?...

Reply:
> Though I guess where do you draw the line?

That's kind of where I'm at.

Also though, we have no code that generates `Baggage` values, nor code that uses them. We're simply passing them along from one place to another. I'd imagine an empty key is unexpected and invalid, but it could be an overreach for us to drop it on the floor when we can easily just pass it along.

Reply:
Yep I agree. I used to do a little open-source on the aws/azure terraform providers...

They would often break because of little things like AWS increasing the number of characters allowed in a field (they duplicated the "rules", and the rules changed, and consumers of the providers only saw the AWS docs).